### PR TITLE
Sign, Modulus, and Range analysis for C++ using sharable semantic layer

### PR DESCRIPTION
**Pull request by dbartol**
This change implements range analysis, sign analysis, and modulus analysis for C++. The vast majority of the implementation is sharable with other languages via a new "semantic layer", which will be described in more detail below. The implementations of the three analyses started from the Java implementations of those analyses.

# Semantic Layer
The Semantic Layer is a library that exposes a program representation suitable for doing semantic analysis on any of our statically-typed languages (C++, C#, Java, Go, Swift). Its usefulness for dynamic languages is unknown. The goal of the semantic layer is to provide a desugared, partial representation of the program's semantics based on a more-or-less fixed set of operations. Each language implements a small, well-defined interface that adapts that language to what the semantic layer expects. The semantic layer itself consumes the language-specific interface and exposes the language-neutral interface, which is in turn consumed by shared analyses.

## Components

The following are the major components of the semantic layer:

### SemType - Type System
A `SemType` represents the type of the result of a `SemExpr`. Unlike the rest of the objects in the semantic layer, types are "folded" together so that types that behave the same share the same identity. For example, integer types are differentiated only by signedness and size. This type system is essentially identical to the type system used by the C++ IR.

### SemExpr - Expressions
In the semantic layer, an "expression" is an operation that consumes zero or more input values and produces a result value. Each expression has a type (`SemType`), an opcode, and an appropriate number of operands, which are themselves expressions. The set of opcodes is shared between languages. Opcodes are defined for fundamental arithmetic and bitwise operators, plus loads, stores, parameter initialization, etc. There is also an `Unknown` opcode to represent an expression whose operation cannot be easily expressed using the official opcodes.

Each language exposes a type, `Expr`, that provides the identity for `SemExpr`. Each instance of `Expr` maps to exactly one `SemExpr`. The language provides the expression's type, opcode, and operand `Expr`s. The semantic layer then exposes a richer interface to the shared analyses. For Java, most Java expressions map to a single `SemExpr` object, but some Java expressions generate multiple `SemExpr` objects. For example, the post-increment (`x++`) operator generates one `SemExpr` to copy the original value of the variable to the result, and a second `SemExpr` to increment the value of the variable afterward. For C++, there's a one-to-one mapping from `Instruction` (in the IR) to `SemExpr`.

### SemSsaVariable - SSA
A `SemSsaVariable` represents a single SSA definition. This may be an explicit definition (such as from an assignment), an implicit definition (such as the initialization of a function parameter), or a Phi definition. The semantic layer assumes that the language has already built an SSA form; the semantic layer _does not_ recompute SSA itself.

### SemBasicBlock - CFG
A  `SemBasicBlock` represents a basic block in the function's control-flow graph (CFG). As with `SemSsaVariable`, it is assumed that the language has already built a CFG.

### SemGuard and SemBound
These two classes are wrappers around the language-specific implementations of guards and bounds.

## Differences from the C++ IR
While the semantic layer looks very similar to the existing C++ IR, there are a few key differences that make the semantic layer easier to implement for a given language:
1. No control-flow threading of individual instructions. Instructions occur in basic blocks, but there is no ordering within a basic block other than that implied by dependencies between a use and its definition.
2. Sparse. Not all operations need be represented. 
3. Much easier to construct, mostly due to the fact that it builds on the CFG and SSA that the language has probably already created.

### CODE REVIEW

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticExprSpecific.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticExprSpecific.qll:16]: Good question. I think that's what I wound up doing for all the other interfaces. Most of the other predicates in the interface between the semantic layer and the language are non-member predicates, so that they aren't visible on the semantic objects that get exposed to the analysis. For `toString()` and `getLocation()`, though, we wind up exposing them anyway, so I don't think there's a need to hide them. Fixed.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticType.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticType.qll:261]: It's not bound, but that's because it could be any `SemNumericType`. The number of unique `SemNumericType`s is ~20 at most, and most languages have only two or three `SemFloatingPointType`s, so the total numer of tuples is still < 100.

I've added a `fromType = any(SemNumericType n)` to be more explicit.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticType.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticType.qll:275]: I don't know about bad RA. I just always get confused about exactly what implication means, so I rarely wind up writing one:smirk:.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeUtils.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeUtils.qll:54]: I think so, but this code is preserved from the original Java implementation so I didn't look too closely.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeUtils.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeUtils.qll:112]: Oh, right. There used to be a disjunction in there to handle the `-=` case, but with the semantic layer, we desugar those.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticExprSpecific.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticExprSpecific.qll:69]: It's quite a bit of work to untangle the parts of the IR's `Opcode` that are needed for the semantic layer from the parts that are IR-specific. It's doable, but I'd prefer to do it in a separate PR.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticType.qll
Comment (jbj) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticType.qll:275]: Using `implies` doesn't let you avoid those parentheses: it binds weaker than `and`. And `P implies Q` should generate exactly the same RA as `not P or Q`.

You may be thinking of the option to write `if P then Q else any()`, which is logically equivalent to `P implies Q` but generates different RA (and has different monotonicity structure).

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeUtils.qll
Comment (jbj) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeUtils.qll:54]: The `booleanXor` version is monotonic in more variables than the `if` version (which may matter if it's recursive), and it may generate better RA.

**Review by MathiasVP — COMMENTED** [review]
I've now looked at all the `cpp/analysis` code as well. I'll take a second pass over it later today or tomorrow.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysis.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysis.qll:462]: As I understand the range analysis, it's OK for a single expression to have many bounds. When a client asks for the bounds of an expression, we attempt to filter out any redundant bounds (e.g., `x > 0 && x < b + 7 && x < b + 9` becomes just `x > 0 && x < b + 7`).

The `Specific::hasBound()` predicate should be able to add bounds even if the core library already computes bounds itself. Either the bounds will be identical, in which case they get folded by QL, or they'll be redundant, in which case they get folded before being presented to the client, or they'll be independent, in which case we want to preserve both bounds.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeUtils.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeUtils.qll:98]: This case is a little different from the range bound case, because I think there should only be a single value of `delta` for a given pair `(e2, e1)`. If the core library and the specific extension disagree on the delta, then one of them is wrong. If the core library is wrong, then it would be wrong even without the extension, because its analysis is based on the well-defined semantics of the expressions it looks at. And if the specific extension is wrong, then it's wrong even if the core library can't figure out a bound.

The overall handling of core analysis vs. specific extension is that the specific extension should only handle cases that require language-specific information, or that depend on language-specific operations (`Opcode::Unknown`). The specific extension can return a _better_ result than the core analysis, but should never return a result that _conflicts with_ the core analysis.

Comment (MathiasVP) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeUtils.qll:98]: Thanks for clearing that up. It sounds like we should have some kind of `RangeAnalysisConsistency` file or similar for this. It's totally not blocking this PR, so I'll mark this as resolved.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysis.qll
Comment (rdmarsh2) [inline cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysis.qll:529]: Do we actually want this to produce two reasons (and two tuples) when `r2` isn't `SemNoReason`

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysis.qll
Comment (rdmarsh2) [inline cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysis.qll:595]: This could use a better explanation

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticExpr.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticExpr.qll:232]: Yes (or more broadly, boxing and unboxing of any value type, including Java primitive types and C# structs).

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/ModulusAnalysis.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/ModulusAnalysis.qll:226]: Oops, that we debugging code. Removed.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysis.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysis.qll:529]: That's the same code that the original Java analysis was using, so I assume it's intended.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/ModulusAnalysis.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/ModulusAnalysis.qll:214]: That was a wrapper that was part of the debugging code from the next comment. It's gone now, and the wrapped predicate does have QLDoc.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticExprSpecific.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticExprSpecific.qll:132]: For the predicates exposed as part of the "signature" to the parameterized module, I've treated them more like IPA constructors, even when they happen to look like simple tests.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticType.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/SemanticType.qll:260]: I'm not sure we're handling infinities correctly, and I know we're not dealing with NaNs at all. This is how the original Java range analysis and shared sign analysis worked, though, so if we want better handling of floating point, we'll probably need to just review the whole semantic sign/range analysis code for FP issues.

File: cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysis.qll
Comment (dbartol) [reply cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysis.qll:595]: I've copied the content from the wrapped predicate.

**Review by MathiasVP — COMMENTED** [review]
LGTM! I'll let @rdmarsh2 do a final review before we merge.

**Review by rdmarsh2 — DISMISSED** [review]
LGTM!

**Comment by rdmarsh2** [comment]
It looks like there's a compilation failure in one of the tests. I'm happy to merge once that's fixed.

**Review by MathiasVP — APPROVED** [review]

**Pull request merged**
